src/cluster_import.c

```diff
@@ -0,0 +1,1948 @@
+ *      │  └─────────────────────────┬─────────────┘                    └──────────┬────────────────┘
+ *      │                            ▼ On PAUSED                                   ▼  If still paused
+ *      │(two phase)┌──────────────────────────────┐                    ┌────────────────────────────┐
+ *      └───────────┤                              ├─REQUEST-FAILOVER──►│                            │
+ *       If denied  │SLOT_IMPORT_FAILOVER_REQUESTED│                    │SLOT_EXPORT_FAILOVER_GRANTED│
+ *      ┌───────────┤                              │◄─FAILOVER-GRANTED/─│                            │
+ *      │(one shot) └────────────────┬─────────────┘  FAILOVER-DENIED   └──────────┬─────────────────┘
+ *      │                            ▼ If granted                                  ▼  On slots transferred
+ *      │             ┌────────────────────────────┐  Cluster bus       ┌───────────────────────────┐
+ *      │             │SLOT_IMPORT_FAILOVER_GRANTED├──broadcast────────►│SLOT_MIGRATION_LINK_SUCCESS│
+ *      │             └──────────────┬─────────────┘                    └───────────────────────────┘
+ *      │                            ▼
+ *      │              ┌───────────────────────────┐
+ *      │              │SLOT_MIGRATION_LINK_SUCCESS│
+ *      │              └───────────────────────────┘
+ *      │
+ *      │               ┌──────────────────────────┐
+ *      └──────────────►│SLOT_MIGRATION_LINK_FAILED│
+ *                      └──────────────────────────┘
+ */
+void proceedWithSlotMigration(slotMigrationLink *link) {
+    /* Continue within the state machine until we have no more work. */
+    while (1) {
+        switch (link->state) {
+        /* Importing states */
+        case SLOT_IMPORT_CONNECTING:
+            if (!link->conn) {
+                connectSlotImportLink(link);
+            }
+            if (proceedWithSlotImportLinkConnecting(link)) {
+                continue;
+            }
+            return;
+        case SLOT_IMPORT_AUTHENTICATING:
+            performSlotImportLinkAuthentication(link);
+            continue;
+        case SLOT_IMPORT_START_SNAPSHOT:
+            initSlotImportLinkClient(link);
+            performSlotImportLinkSnapshot(link);
+            return;
+        case SLOT_IMPORT_FAILOVER_GRANTED:
+            performSlotImportLinkFailover(link);
+            return;
+        case SLOT_IMPORT_FINISHED_WAITING_TO_CLEANUP:
+            delKeysNotOwnedByMyself(link->slot_ranges);
+            updateSlotMigrationLinkState(link, link->post_cleanup_state);
+            return;
+        case SLOT_IMPORT_RECEIVE_SNAPSHOT:
+        case SLOT_IMPORT_RECEIVE_STREAM:
+        case SLOT_IMPORT_FAILOVER_WAITING_FOR_PAUSED:
+        case SLOT_IMPORT_FAILOVER_REQUESTED:
+            return;
+
+        /* Exporting states */
+        case SLOT_EXPORT_WAITING_TO_SNAPSHOT:
+            /* Perform the snapshot whenever the opportunity arises. */
+            if (hasActiveChildProcess()) {
```

Comment:
At this point in time we are already setup to collect commands for the stream, correct? (If I get this correctly, this happens in `SLOT_EXPORT_WAITING_TO_SNAPSHOT` as soon as the export link is configured).

The stream must start exactly at the state captured by the snapshot. But, if we postpone the snapshot here (to some later cron execution?), how do we synchronize the two? I can't find code to do that (I'd expect code here to start recording the stream)

Reply:
Good point. We need to ensure we skip feeding slot migrations if they are in SLOT_EXPORT_WAITING_TO_SNAPSHOT (only after it enters SLOT_EXPORT_SNAPSHOTTING). Otherwise we will duplicate some commands